### PR TITLE
Fix our use of suppressParent in project.json

### PR DESCRIPTION
**Pull request by jaredpar**
Infrastructure change

TLDR: This change removes our use of suppressParent which should not be used

The core problem in our build setup which lead to the use of suppressParent is having
the same reference DLL introduced by different NuGet packages.  These packages differed
in name and version from each other.  The most notable example is the reference DLL for Microsoft.VisualStudio.Language.NavigateTo.Interfaces.dll which is contained in 3
different NuGet packages:

- RoslynDependencies.Microsoft.VisualStudio.Language.NavigateTo.Interfaces 14.0
- Roslyn.Microsoft.VisualStudio.Language.NavigateTo.Interfaces 15.0 preview 5
- Microsoft.VisualStudio.Language.NavigateTo.Interfaces 15.0 RC

The differing names is a problem because when resolving conflicts NuGet doesn't consider
the referenced DLL versions at all (by design).  Instead it is only concerned with
handling conflicts between packages of the same name.

These packages have different name and hence NuGet never attempts to do any conflict
resolution between them.  It will consider each package to be a separate entity and
pass on their assets to MSBuild.  This means that MSBuild will eventually be handed the
same DLL with 3 different versions and consequently begins issuing MSB3277 errors.

The suppressParent entries in the project.json file suppressed this error because it
essentially removes the listed package from the transitive graph.  Hence it never appeared
in referenced projects, only a single DLL was passed to MSBuild and compilations
progressed.

This is the wrong approach to fixing that problem because it's subverting both the
depenedncy conflict resolution aspects of NuGet / MSBuild and causing us to create
incomplete deployments in our unit test directories.  This is fighting the tooling
instead of leveraging it.

The more robust approach to solving this problem is to have a reference DLL always
distributed through the same NuGet package.  This allows NuGet to handle the version
conflicts using standard conflict rules and resulting in only a single DLL being passed
to MSBuild.

In the past this has been a blocker because we often need DLLs at versions that aren't
available on NuGet.  Going forward we will be working with the VSSDK to remedy that
problem.  Short term though we are simply going to upload ad-hoc packages with the correct
name to the roslyn-tools feed using the pre-release moniker -alpha.  This ensures we
don't have any conflicts with official packages on NuGet.org.

There are a few cases this change doesn't completely address that I want to call
out:

- Types moved between the MS.VS.Shell.Immutable and MS.VS.Shell.Framework DLLs between
  Dev14 and Dev15.  To prevent a lot of duplicate type errors the MS.VS.Shell.Immutable
  DLLs need to be removed from the compile graph, but not the runtime graph, in our
  Next projects.
- GraphModel is an adhoc package created by us that doesn't have an existing NuGet
  package to pattern off of and it's not obvious how such a package would be
  laid out if it existed.  The Dev15 packages also include this as a reference by
  default which causes a confilct with our packages.  As such I've used "include: none"
  for now to work around the problem until a final NuGet package is decided on.

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @Pilchie, @dotnet/roslyn-infrastructure

**Comment by jaredpar** [comment]
@Pilchie 

> Love it once the issue in Xaml build is fixed. 

Looking into this now.  It's only an issue when building from a machine with Dev15 installed.  That adds new reference probing paths and creates a couple of new conflicts.  Should have a fix here in a bit. 

> Will this mean that @CyrusNajmabadi can finally write unit tests for Dev15 functionality that depend on Dev15 types?

Yes.  But that should also be possible today in the Next projects.  What was blocking it before?

**Comment by CyrusNajmabadi** [comment]
> Yes. But that should also be possible today in the Next projects. What was blocking it before?

Yes.  These are the issues i've pinged you about a few times in the last few months.  I've been blocked on testing Dev15 features.  Just trying to reference the right VS binaries with the APIs i needed would make testing break.

**Comment by jaredpar** [comment]
@CyrusNajmabadi 

> Just trying to reference the right VS binaries with the APIs i needed would make testing break.

It's unclear what breaks though.  Does it mean you couldn't restore, couldn't build, couldn't get MSBuild to stop issuing conflicts, or the tests when run on Jenkins failed due to missing binaries?  

I'm further confused because we are running Dev15 tests today from what I can see.  I checked the most recent log and  Roslyn.VisualStudio.Next.UnitTests.dll is indeed executing and running tests.  I assume that's using Dev15 APIs based on the project references.  

Things I definitely think are a problem in this area:

1. Changing the existing test libraries to reference Dev15.  That should be easier after this change.  It's been possible since the build was broken up but I imagine "head to keyboard" frustrating to make work.  My head certainly banged my keyboard a few times getting this change in 😦   
2. Knowing what Dev15 version to reference in our code base.  This is highly dependent on what our Jenkins machines have installed.  @tannergooding or @tmeschter should know.  But really we need to change our Jenkins machine labels to be absolutely explicit about what Dev15 version they have.  Otherwise it's guesswork and easy to have changes run locally and fail on CI.  
3. Not having the right VS SDK reference to use in NuGet.  

If the issue is 1 I can certainly help out here (once I have the answers to 2).

**Comment by jaredpar** [comment]
Should add that my plan is to move them all over to Dev15 myself.  I've been blocked from doing so fully by the roslyn-internal changes ( @tmeschter is driving).  Once that is done though I plan to bulk move the rest of our code to Dev15 references very quickly.

**Comment by CyrusNajmabadi** [comment]
@jaredpar see https://github.com/dotnet/roslyn/pull/14283 as an example of this.  Looks like it's now failing with:

MESSAGE:
System.IO.FileLoadException : Could not load file or assembly 'Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
+++++++++++++++++++

**Comment by jaredpar** [comment]
@CyrusNajmabadi I don't see how that failure is related to Dev15 at all.  It looks like the code is now using Newtonsoft.Json and has a bad binding redirect.  It's loading 6.0.0.0, from where I don't know, but all of our references are in the 8.0.0.0 range (mandated by VS).

**Comment by jaredpar** [comment]
@CyrusNajmabadi i confirmed this has nothing to do with Dev15.  This a problem in the StreamJsonRpc library.  It references Newtonsoft.Json 6.0 while the entirety of VS uses 8.0 including all Roslyn references.  

To fix this both of the following config files need to get a binding redirect for Newtonsoft.Json from 6.0.0.0 to 8.0.3.0:

- https://github.com/dotnet/roslyn/blob/master/build/Defaults/Portable/app.config
- https://github.com/dotnet/roslyn/blob/master/build/Defaults/Desktop/app.config

**Comment by jasonmalinowski** [comment]
@jaredpar Late, but thanks a million for cleaning this mess up, and also for writing a good write-up of the situation I can point people to the next time this is about to get broken again.

**Comment by heejaechang** [comment]
@jaredpar so with this, can I move to new StreamJsonRpc nuget which uses newer versino of newtonsoft.Json and dev15 dependencies?

**Comment by jaredpar** [comment]
@heejaechang No. This PR just fixes our dependency tree to rational. It doesn't move us to Dev15. That is what must happen before we can move StreamJsonRpc forward. 

See my most recent PR. That is what moves us to Dev15. As part of that change I also moved StreamJsonRpc forward.

**Comment by heejaechang** [comment]
cool. thank you.
